birdvoxpaint/util.py

```diff
@@ -0,0 +1,225 @@
+#             y_complete, _ = librosa.load(filename, duration=offset)
+#             remaining = np.concatenate([remaining, y_complete])
+#
+#             if len(remaining) == block_length: # there were enough samples in the new file
+#                 yield remaining
+#                 remaining = None # clear
+#             else: # super short file ??? - better to be robust to all types of weather.
+#                 continue
+#
+#         y_blocks = librosa.stream(filename,
+#             offset=offset,
+#             block_length=block_n_frames,
+#             frame_length=frame_length,
+#             hop_length=hop_length)
+#
+#         for y in y_blocks:
+#             librosa.util.valid_audio(y, mono=True)
+#             if sr != orig_sr:
+#                 librosa.resample(y, orig_sr, sr)
+#             yield y
+
+
+
+def spec_frame(S, segment_length):
+    '''Break a spectrogram into equal segments.
+
+    Returns:
+        Spectrograms with the window applied
+        shape: (freq, frame * time) => (frame, freq, time)
+
+    '''
+
+    # only feed in full length frames into the pipeline
+    truncated_length = (S.shape[1] // segment_length) * segment_length
+    S = S[:,:truncated_length] # removes incomplete frames
+    S = S.T.reshape(-1, segment_length, len(S)) # shape: frame, segment, freq
+    S = S.transpose((0, 2, 1)) # shape: frame, freq, segment
+    return S
+
+
+
+def apply_indices(S, indices, segment_length, n_jobs=1):
+    '''Takes an iterable yielding blocks of spectrograms, breaks them into blocks of
+    size: segment_length. The number of blocks per yield should equal n_jobs.
+
+    Arguments:
+        S (iterable(np.ndarray[freq, time])): the iterable of spectrograms
+        indices (callable/list of callables): the indices to calculate. If a
+            single callable, it should return all indices as the final axis.
+        segment_length (int): the length in time (# frames) to be passed to
+            indices at a time.
+
+    '''
+    # reshape to blocks of spectrograms (# blocks == n_jobs)
+    S_blocks = (spec_frame(spec, segment_length) for spec in S) # yields: block, freq, time
+
+    # allow user to pass a custom indices function - or as a list of functions
+    if callable(indices):
+        calc_indices = indices
+    else:
+        calc_indices = lambda x: np.stack([f(x) for f in indices], axis=-1)
+
+    # build job to calculate indices
+    if n_jobs == 1:
+        jobs = joblib.Parallel(n_jobs=n_jobs)
+        calc_indices = joblib.delayed(calc_indices)
+    else:
+        jobs = lambda x: x # dummy function so that we can call jobs below like normal.
+
+    S_blocks = (
+        np.stack(list(jobs(map(calc_indices, spec))), axis=1) # shape: freq, block
+        for spec in S_blocks # yields: block, freq, time
+        if spec.size # removes empty blocks
+    )
+
+    S = np.concatenate(list(S_blocks), axis=1) # shape: freq, block, indices
+    return S
+
+def _shape(X):
```

Comment:
unused function

Reply:
`_shape` was just for debugging in places where I wanted to probe the shape inside a generator comprehension and it seemed like the easiest way to do it. can totally delete when we don't need it anymore